File: manda-processing/src/observability/usage.py
"""
Usage logging service for LLM and feature tracking.
Story: E12.2 - Usage Logging Integration (AC: #1, #4, #5)

This module provides async functions to persist usage data to PostgreSQL
using asyncpg (following the established supabase_client.py pattern).
"""

import json
import asyncpg
import structlog
from typing import Any, Optional
from uuid import UUID

from src.storage.supabase_client import SupabaseClient

logger = structlog.get_logger(__name__)


async def log_llm_usage_to_db(
    db: SupabaseClient,
    *,
    organization_id: Optional[UUID] = None,
    deal_id: Optional[UUID] = None,
    user_id: Optional[UUID] = None,
    provider: str,
    model: str,
    feature: str,
    input_tokens: int,
    output_tokens: int,
    cost_usd: float,
    latency_ms: Optional[int] = None,
) -> Optional[str]:
    """
    Persist LLM usage to database using asyncpg.

    IMPORTANT: Uses asyncpg raw SQL pattern (NOT Supabase SDK).
    See supabase_client.py for reference pattern.

    Args:
        db: SupabaseClient instance (provides asyncpg pool)
        organization_id: Organization for multi-tenant isolation (E12.9)
        deal_id: Deal context for the LLM call
        user_id: User who initiated the call
        provider: LLM provider (google-gla, anthropic, voyage, openai)
        model: Model identifier (gemini-2.5-flash, claude-sonnet-4-0)
        feature: Feature using LLM (chat, document_analysis, extraction, etc.)
        input_tokens: Number of input tokens
        output_tokens: Number of output tokens
        cost_usd: Calculated cost in USD
        latency_ms: Optional latency in milliseconds

    Returns:
        Created record ID as string, or None if insert failed
    """
    try:
        pool = await db._get_pool()
        async with pool.acquire() as conn:
            row = await conn.fetchrow(
                """
                INSERT INTO llm_usage (
                    organization_id, deal_id, user_id,
                    provider, model, feature,
                    input_tokens, output_tokens, cost_usd, latency_ms
                ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
                RETURNING id
                """,
                organization_id,
                deal_id,
                user_id,
                provider,
                model,
                feature,
                input_tokens,
                output_tokens,
                round(cost_usd, 6),
                latency_ms,
            )
            return str(row["id"]) if row else None

    except asyncpg.PostgresError as e:
        logger.error(
            "llm_usage_db_insert_failed",
            error=str(e),
            provider=provider,
            model=model,
            feature=feature,
        )
        return None
    except Exception as e:
        logger.error(
            "llm_usage_unexpected_error",
            error=str(e),
            provider=provider,
            model=model,
        )
        return None


async def log_feature_usage_to_db(
    db: SupabaseClient,
    *,
    organization_id: Optional[UUID] = None,
    deal_id: Optional[UUID] = None,
    user_id: Optional[UUID] = None,
    feature_name: str,
    status: str,
    duration_ms: Optional[int] = None,
    error_message: Optional[str] = None,
    metadata: Optional[dict] = None,
) -> Optional[str]:
    """
    Persist feature usage to database using asyncpg.

    IMPORTANT: Uses asyncpg raw SQL pattern (NOT Supabase SDK).

    Args:
        db: SupabaseClient instance (provides asyncpg pool)
        organization_id: Organization for multi-tenant isolation (E12.9)
        deal_id: Deal context for the feature execution
        user_id: User who initiated the feature
        feature_name: Name of feature (upload_document, chat, search, etc.)
        status: Execution status (success, error, timeout)
        duration_ms: Optional execution duration in milliseconds
        error_message: Error message if status is 'error'
        metadata: Optional additional context as JSON

    Returns:
        Created record ID as string, or None if insert failed
    """
    try:
        # Convert metadata dict to JSON string for JSONB column
        metadata_json = json.dumps(metadata) if metadata else None

        pool = await db._get_pool()
        async with pool.acquire() as conn:
            row = await conn.fetchrow(
                """
                INSERT INTO feature_usage (
                    organization_id, deal_id, user_id,
                    feature_name, status, duration_ms,
                    error_message, metadata
                ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8::jsonb)
                RETURNING id
                """,
                organization_id,
                deal_id,
                user_id,
                feature_name,
                status,
                duration_ms,
                error_message,
                metadata_json,
            )
            return str(row["id"]) if row else None

    except asyncpg.PostgresError as e:
        logger.error(
            "feature_usage_db_insert_failed",
            error=str(e),
            feature_name=feature_name,
            status=status,
        )
        return None
    except Exception as e:
        logger.error(
            "feature_usage_unexpected_error",
            error=str(e),
            feature_name=feature_name,
        )
        return None

File: manda-app/lib/observability/usage.ts
/**
 * Usage logging service for LLM and feature tracking.
 * Story: E12.2 - Usage Logging Integration (AC: #2, #4, #5)
 *
 * This module provides functions to persist usage data to Supabase,
 * enabling cost visibility and performance analysis in E12.3 dashboard.
 */

import { createClient } from '@/lib/supabase/server'
import type { Database } from '@/lib/supabase/database.types'

type LlmUsageInsert = Database['public']['Tables']['llm_usage']['Insert']
type FeatureUsageInsert = Database['public']['Tables']['feature_usage']['Insert']

/**
 * Log LLM usage to database.
 *
 * @param params - LLM usage parameters
 * @returns Created record ID, or null if insert failed
 *
 * @example
 * ```typescript
 * const startTime = Date.now()
 * const result = await llm.invoke(messages)
 * await logLLMUsage({
 *   dealId: state.dealId,
 *   userId: state.userId,
 *   organizationId: state.organizationId,
 *   provider: 'anthropic',
 *   model: 'claude-sonnet-4-0',
 *   feature: 'chat',
 *   inputTokens: result.usage.input_tokens,
 *   outputTokens: result.usage.output_tokens,
 *   costUsd: calculateCost(result.usage),
 *   latencyMs: Date.now() - startTime,
 * })
 * ```
 */
export async function logLLMUsage(params: {
  organizationId?: string
  dealId?: string
  userId?: string
  provider: string
  model: string
  feature: string
  inputTokens: number
  outputTokens: number
  costUsd: number
  latencyMs?: number
}): Promise<string | null> {
  try {
    const supabase = await createClient()

    const insertData: LlmUsageInsert = {
      provider: params.provider,
      model: params.model,
      feature: params.feature,
      input_tokens: params.inputTokens,
      output_tokens: params.outputTokens,
      cost_usd: params.costUsd,
    }

    // Add optional fields
    if (params.organizationId) {
      insertData.organization_id = params.organizationId
    }
    if (params.dealId) {
      insertData.deal_id = params.dealId
    }
    if (params.userId) {
      insertData.user_id = params.userId
    }
    if (params.latencyMs !== undefined) {
      insertData.latency_ms = params.latencyMs
    }

    const { data, error } = await supabase
      .from('llm_usage')
      .insert(insertData)
      .select('id')
      .single()

    if (error) {
      console.error('[logLLMUsage] Insert failed:', error.message)
      return null
    }

    return data?.id ?? null
  } catch (error) {
    console.error('[logLLMUsage] Error:', error)
    return null
  }
}

/**
 * Log feature usage to database.
 *
 * @param params - Feature usage parameters
 * @returns Created record ID, or null if insert failed
 *
 * @example
 * ```typescript
 * const startTime = Date.now()
 * try {
 *   await processDocument(doc)
 *   await logFeatureUsage({
 *     dealId: doc.dealId,
 *     userId: uploaderId,
 *     organizationId: orgId,
 *     featureName: 'document_upload',
 *     status: 'success',
 *     durationMs: Date.now() - startTime,
 *     metadata: { documentCount: 1, totalBytes: doc.size },
 *   })
 * } catch (err) {
 *   await logFeatureUsage({
 *     dealId: doc.dealId,
 *     featureName: 'document_upload',
 *     status: 'error',
 *     durationMs: Date.now() - startTime,
 *     errorMessage: err.message,
 *     metadata: { stack: err.stack },
 *   })
 * }
 * ```
 */
export async function logFeatureUsage(params: {
  organizationId?: string
  dealId?: string
  userId?: string
  featureName: string
  status: 'success' | 'error' | 'timeout'
  durationMs?: number
  errorMessage?: string
  metadata?: Record<string, unknown>
}): Promise<string | null> {
  try {
    const supabase = await createClient()

    const insertData: FeatureUsageInsert = {
      feature_name: params.featureName,
      status: params.status,
    }

    // Add optional fields
    if (params.organizationId) {
      insertData.organization_id = params.organizationId
    }
    if (params.dealId) {
      insertData.deal_id = params.dealId
    }
    if (params.userId) {
      insertData.user_id = params.userId
    }
    if (params.durationMs !== undefined) {
      insertData.duration_ms = params.durationMs
    }
    if (params.errorMessage) {
      insertData.error_message = params.errorMessage
    }
    if (params.metadata) {
      insertData.metadata = params.metadata as Database['public']['Tables']['feature_usage']['Insert']['metadata']
    }

    const { data, error } = await supabase
      .from('feature_usage')
      .insert(insertData)
      .select('id')
      .single()

    if (error) {
      console.error('[logFeatureUsage] Insert failed:', error.message)
      return null
    }

    return data?.id ?? null
  } catch (error) {
    console.error('[logFeatureUsage] Error:', error)
    return null
  }
}

/**
 * Helper to calculate LLM cost from token counts.
 *
 * Pricing (per 1M tokens):
 * - Gemini 2.5 Flash: $0.075 input, $0.30 output
 * - Claude Sonnet 4: $3.00 input, $15.00 output
 * - Voyage 3.5: $0.06 input, $0 output (embeddings)
 */
export function calculateLLMCost(
  provider: string,
  model: string,
  inputTokens: number,
  outputTokens: number
): number {
  // Pricing per 1M tokens
  const pricing: Record<string, { input: number; output: number }> = {
    'gemini-2.5-flash': { input: 0.075, output: 0.30 },
    'gemini-2.5-pro': { input: 1.25, output: 5.00 },
    'claude-sonnet-4-0': { input: 3.00, output: 15.00 },
    'claude-opus-4-5': { input: 15.00, output: 75.00 },
    'voyage-3.5': { input: 0.06, output: 0 },
    'rerank-2.5': { input: 0.05, output: 0 },
  }

  const rates = pricing[model] ?? { input: 0, output: 0 }
  return (
    (inputTokens * rates.input) / 1_000_000 +
    (outputTokens * rates.output) / 1_000_000
  )
}
